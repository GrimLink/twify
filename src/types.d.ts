import { drivers } from './drivers';

export interface Step {
  (): Promise<void>;
}

export interface Framework {
  requiredDependencies: string[];
  initCommands: string[];
  cssLocation: string;
  content: {
    name: string;
    files: string[];
  };
  steps: Step[];
}

export type DriverImport = Promise<{ default: Framework }>;

export type Driver = keyof typeof drivers;
