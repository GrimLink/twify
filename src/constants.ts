export const CSS_STUB = `@tailwind base;
@tailwind components;
@tailwind utilities;
`;
