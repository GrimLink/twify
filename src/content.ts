import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import j from 'jscodeshift';
import { Framework } from './types';

export function addContentToCode(code: string, content: string[]): string {
  const parse = j.withParser('flow');
  const ast = parse(code);

  ast
    .find(j.Property, { key: { name: 'content' } })
    .replaceWith(
      j.property(
        'init',
        j.identifier('content'),
        j.arrayExpression(content.map((c) => j.literal(c)))
      )
    );

  return ast.toSource();
}

export async function setupContent({ content }: Framework) {
  console.log(`\nConfiguring ${chalk.blue.bold(content.name)} content...`);
  console.log(chalk.blue(`- ${content.files.join('\n- ')}`));

  const contentPath = path.join(process.cwd(), content.name);

  const fileContent = await fs.readFile(contentPath, 'utf8');
  const newContent = addContentToCode(fileContent, content.files);

  await fs.writeFile(contentPath, newContent);
}
