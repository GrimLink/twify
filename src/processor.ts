import fs from 'fs-extra';
import chalk from 'chalk';
import { Framework } from './types';
import { CSS_STUB } from './constants';
import { detectInstaller, runCommand } from './helpers';
import { setupContent } from './content';

export async function handle(framework: Framework) {
  const { requiredDependencies, initCommands, cssLocation, steps } = framework;
  const installer = detectInstaller();

  // Install required dependencies
  if (requiredDependencies.length) {
    console.log(`\nInstalling Required Dependencies...`);
    console.log(chalk.blue(`- ${requiredDependencies.join('\n- ')}`));
    await runCommand(`${installer} ${requiredDependencies.join(' ')}`);
  }

  // Run init commands
  if (initCommands.length) {
    console.log(`\nInitializing Config Files...`);
    for (const command of initCommands) {
      await runCommand(command);
    }
  }

  // Write content to cssLocation
  console.log(`\nSetting up ${chalk.blue.bold(cssLocation)}...`);
  fs.ensureFileSync(cssLocation);
  const exitingCss = fs.readFileSync(cssLocation, 'utf8');
  fs.writeFileSync(cssLocation, `${exitingCss}\n\n${CSS_STUB}`);

  await setupContent(framework);

  if (steps.length) {
    // Run process
    console.log(`\nRunning ${steps.length} step(s)...`);
    for (const stepFunc of steps) {
      await stepFunc();
    }
  }
}
