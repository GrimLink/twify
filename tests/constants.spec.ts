import { CSS_STUB } from '../src/constants';

describe('Constants', () => {
  it('has default css content', () => {
    expect(CSS_STUB).toMatchSnapshot();
  });
});
