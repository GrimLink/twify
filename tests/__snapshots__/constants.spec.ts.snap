// Vitest Snapshot v1

exports[`Constants > has default css content 1`] = `
"@tailwind base;
@tailwind components;
@tailwind utilities;
"
`;
